File: Python/ByRound/1981/1981_E_Turtle_and_Intersected_Segments.py
# Problem: CF 1981 E - Turtle and Intersected Segments
# https://codeforces.com/contest/1981/problem/E

"""
Algorithm: Union-Find with Segment Tree for Minimum Spanning Tree
Time Complexity: O(n log n)
Space Complexity: O(n)
This solution uses a segment tree to efficiently manage intervals and a union-find structure
to build the minimum spanning tree. The key idea is to sort segments by their 'a' values
and use the segment tree to track which intervals intersect with the current one.
"""

import io
import os
import sys

input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline


def f(u, v):
    # Helper function to encode two values into one
    return u << 20 ^ v


def update(i, x, y):
    # Update segment tree at position i with min x and max y
    i ^= l1
    mi[i], ma[i] = x, y
    i >>= 1
    while i:
        mi[i] = min(mi[i << 1], mi[i << 1 ^ 1])
        ma[i] = max(ma[i << 1], ma[i << 1 ^ 1])
        i >>= 1
    return


def get_min(s, t):
    # Get minimum value in range [s,t] from segment tree
    s, t = s ^ l1, t ^ l1
    ans = inf
    while s <= t:
        if s & 1:
            ans = min(ans, mi[s])
            s += 1
        s >>= 1
        if not t & 1:
            ans = min(ans, mi[t])
            t -= 1
        t >>= 1
    return ans


def get_max(s, t):
    # Get maximum value in range [s,t] from segment tree
    s, t = s ^ l1, t ^ l1
    ans = -inf
    while s <= t:
        if s & 1:
            ans = max(ans, ma[s])
            s += 1
        s >>= 1
        if not t & 1:
            ans = max(ans, ma[t])
            t -= 1
        t >>= 1
    return ans


def get_root(s):
    # Find root of s in union-find structure
    while s ^ root[s]:
        s = root[s]
    return s


def unite(s, t):
    # Union two sets in union-find structure
    s, t = get_root(s), get_root(t)
    if s == t:
        return
    if rank[s] < rank[t]:
        s, t = t, s
    if rank[s] == rank[t]:
        rank[s] += 1
    root[t] = s
    return


def same(s, t):
    # Check if two elements are in same set
    return True if get_root(s) == get_root(t) else False


t = int(input())
ans = []
inf = pow(10, 9) + 1
for _ in range(t):
    n = int(input())
    l, r, a = [0] * n, [0] * n, [0] * n
    for i in range(n):
        l0, r0, a0 = map(int, input().split())
        l[i], r[i], a[i] = l0, r0, f(a0, i)
    u, c = [0] * (n << 1), 0
    a.sort()
    for i in range(n):
        j, a[i] = a[i] & 0xFFFFF, a[i] >> 20
        u[j << 1], u[j << 1 ^ 1] = f(l[j] << 1, i), f(r[j] << 1 ^ 1, i)
    l1 = pow(2, (n + 1).bit_length())
    l2 = 2 * l1
    mi, ma = [inf] * l2, [-inf] * l2
    x, y, c = [0] * (n << 1), [0] * (n << 1), 0
    for i in sorted(u):
        j = i & 0xFFFFF
        if mi[j ^ l1] == inf:
            l, r = get_max(0, j), get_min(j, l1 - 1)
            if l ^ -inf:
                x[c], y[c], c = f(l, j), f(a[j] - a[l], c), c + 1
            if r ^ inf:
                x[c], y[c], c = f(j, r), f(a[r] - a[j], c), c + 1
            update(j, j, j)
        else:
            update(j, inf, -inf)
    root = [i for i in range(n)]
    rank = [1 for _ in range(n)]
    ans0, cnt = 0, 0
    for i in sorted(y[:c]):
        j = i & 0xFFFFF
        l, r = x[j] >> 20, x[j] & 0xFFFFF
        if not same(l, r):
            unite(l, r)
            ans0 += i >> 20
            cnt += 1
        if cnt == n - 1:
            break
    if cnt ^ (n - 1):
        ans0 = -1
    ans.append(ans0)
sys.stdout.write("\n".join(map(str, ans)))


# https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/E/1981_E_Turtle_and_Intersected_Segments.cpp
// Problem: CF 1981 E - Turtle and Intersected Segments
// https://codeforces.com/contest/1981/problem/E

/*
 * Problem: Turtle and Intersected Segments
 * 
 * Algorithm: 
 * - Use a sweep line technique with a map to efficiently find intersecting segments.
 * - For each segment, determine which previous segments it intersects with.
 * - Construct a minimum spanning tree (MST) using Kruskal's algorithm on the resulting edges.
 * 
 * Time Complexity: O(n log n) per test case
 * Space Complexity: O(n)
 * 
 * Key techniques:
 * - Coordinate compression using map for efficient intersection detection
 * - Union-Find (Disjoint Set Union) for MST construction
 * - Sweep line algorithm to reduce the number of intersection checks
 * 
 * Approach:
 * 1. Sort segments by left endpoint
 * 2. Process segments in order, maintaining a map of (a_i, {right_endpoint, index})
 * 3. For each new segment, find intersecting previous segments by checking overlapping intervals
 * 4. Build edges with weights |a_i - a_j| for intersecting segments
 * 5. Apply Kruskal's algorithm to find MST
 * 6. If the graph is not connected, return -1
 */

#include <algorithm>
#include <iostream>
#include <iterator>
#include <map>
#include <numeric>
#include <tuple>
#include <utility>
#include <vector>

using namespace std;
using ll = long long;
using pii = pair<int, int>;
using tii = tuple<int, int, int>;
constexpr int maxn = 500001;
tii d[maxn];
int p[maxn];

// Find root of node with path compression
int find(int node) { return node == p[node] ? node : p[node] = find(p[node]); }

void solve() {
  int n, a, b, c;
  cin >> n;
  for (int i = 0; i < n; i++) {
    cin >> a >> b >> c;
    d[i] = {a, b, c};
  }
  sort(d, d + n); // Sort segments by left endpoint
  iota(p, p + n, 0); // Initialize parent array for Union-Find
  ll ans = 0;
  map<int, pii> mp{{get<2>(d[0]), {get<1>(d[0]), 0}}}; // Map of (a_i, {right_endpoint, index})
  vector<tii> edges; // Store edges: (weight, u, v)
  
  for (int i = 1; i < n; i++) {
    tie(a, b, c) = d[i]; // Current segment (a, b, c)
    
    // Remove segments from map that no longer intersect current segment
    auto p_it = mp.lower_bound(c);
    while (p_it != mp.end() && p_it->second.first < a)
      p_it = mp.erase(p_it);
    
    // Check intersection with segments in the map
    if (p_it != mp.end())
      edges.push_back({p_it->first - c, p_it->second.second, i});
    
    // Check in reverse direction for intersection
    if (p_it != mp.begin()) {
      --p_it;
      while (p_it->second.first < a) {
        p_it = mp.erase(p_it);
        if (p_it != mp.begin())
          p_it = prev(p_it);
        else
          break;
      }
      if (p_it != mp.end() && p_it->first < c)
        edges.push_back({c - p_it->first, p_it->second.second, i});
    }
    
    // Update map with current segment
    if (mp[c].first < b)
      mp[c] = {b, i};
  }
  
  sort(edges.begin(), edges.end()); // Sort edges by weight
  
  int ct = 1; // Count of connected components
  for (auto [c, a, b] : edges) {
    a = find(a), b = find(b); // Find roots
    if (a != b) {
      p[a] = b; // Union
      ct++;
      ans += c; // Add weight to MST
    }
  }
  
  // If all nodes are connected, output sum of weights, otherwise -1
  cout << (ct == n ? ans : -1) << '\n';
}

int main() {
  cin.tie(0)->sync_with_stdio(0);
  int t;
  cin >> t;
  while (t--)
    solve();
  return 0;
}


// https://github.com/VaHiX/CodeForces/